extmod/nimble/hal/hal_uart.c

```diff
@@ -86,15 +86,37 @@ int hal_uart_close(uint32_t port) {
     return 0; // success
 }
 
+STATIC void mp_bluetooth_hci_uart_readchar_cb(uint8_t chr) {
+    hal_uart_rx_cb(hal_uart_rx_arg, chr);
+}
+
+MP_WEAK int mp_bluetooth_hci_uart_readchar(void) {
+    // dummy function in case port only uses mp_bluetooth_hci_uart_readpacket
+    return -2;
+}
+
+MP_WEAK int mp_bluetooth_hci_uart_readpacket(mp_bluetooth_hci_uart_readchar_t handler) {
+    // dummy function in case port only uses mp_bluetooth_hci_uart_readpacket
+    return -2;
+}
+
 void mp_bluetooth_nimble_hci_uart_process(bool run_events) {
     bool host_wake = mp_bluetooth_hci_controller_woken();
 
+    while (mp_bluetooth_hci_uart_readpacket(&mp_bluetooth_hci_uart_readchar_cb) >= 0) {
```

Comment:
I was initially confused why there's two loops here, but it's now clear that they're mutually exclusive. I got a bit nervous about seeing an extra call to `mp_bluetooth_nimble_os_eventq_run_all`.

Would prefer to make that clear:

```
#if MICROPY_PY_BLUETOOTH_HCI_READ_MODE == MICROPY_PY_BLUETOOTH_HCI_READ_MODE_BYTE
int chr;
    while ((chr = mp_bluetooth_hci_uart_readchar()) >= 0) {
        mp_bluetooth_hci_uart_char_cb(chr);
#elif MICROPY_PY_BLUETOOTH_HCI_READ_MODE == MICROPY_PY_BLUETOOTH_HCI_READ_MODE_PACKET
while (mp_bluetooth_hci_uart_readpacket(&mp_bluetooth_hci_uart_char_cb) >= 0) {
#else
#error ...
#endif

// Incoming data may result in events being enqueued. If we're in
        // scheduler context then we can run those events immediately.
        if (run_events) {
            mp_bluetooth_nimble_os_eventq_run_all();
        }
    }
```

Reply:
That's a good point to remove the duplication / make it clear they're separate code paths. I'm not a fan of #defines around while / loop entry points as I have a hard time reading them, if nothing else the preproc vs code indentation is always a bit weird. As such I've tried to do this in a slightly clearer (?) way that hopefully doesn't add any real overhead, see what you think.